### PR TITLE
🐦‍🔥 6x faster GRPO with multi-step optimization

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

This PR implements the multi-step trick. It allows the generation to be reused several times, speeding up training.

It requires to implement the importance sampling and the clipping logic.

```diff
- per_token_loss = -torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
+ coef_1 = torch.exp(per_token_logps - old_per_token_logps)
+ coef_2 = torch.clamp(coef_1, 1 - self.epsilon, 1 + self.epsilon)
+ per_token_loss1 = coef_1 * advantages.unsqueeze(1)
+ per_token_loss2 = coef_2 * advantages.unsqueeze(1)
+ per_token_loss = -torch.min(per_token_loss1, per_token_loss2)
```

Results with various mu (note that mu=1 matches the current implementation in main):

W B Chart 20_02_2025, 13_24_19

W B Chart 20_02_2025, 13_17_55-3

W B Chart 20_02_2025, 13_17_55-2

W B Chart 20_02_2025, 13_18_56-2

For reproduction:

```python
from datasets import load_dataset
from trl import GRPOConfig, GRPOTrainer


dataset = load_dataset("trl-lib/tldr", split="train")

# Dummy reward function: the closer the completion is to 20 characters, the higher the reward
def reward_len(completions, **kwargs):
    return [-abs(100 - len(completion)) for completion in completions]

def main():
    num_iterations=4
    training_args = GRPOConfig(
        output_dir=f"Qwen2.5-0.5B-GRPO-2899-μ={num_iterations}",
        logging_steps=5,
        gradient_accumulation_steps=4,
        per_device_train_batch_size=4,
        num_generations=8,
        max_prompt_length=64,
        max_completion_length=64,
        log_completions=True,
        max_steps=200,
        num_iterations=num_iterations,
    )
    trainer = GRPOTrainer(
        model="Qwen/Qwen2.5-0.5B",
        reward_funcs=reward_len,
        args=training_args,
        train_dataset=dataset,
    )
    trainer.train()


if __name__ == "__main__":
    main()
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_config.py [hunk 1]
```diff
@@ -224,6 +228,14 @@ class GRPOConfig(TrainingArguments):
             "training speed."
         },
     )
+    num_iterations: int = field(
+        default=1,
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py:232]: Should we set another default? eg 4?

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -638,6 +731,10 @@ def _prepare_inputs(self, inputs: dict[str, Union[torch.Tensor, Any]]) -> dict[s
         logits_to_keep = completion_ids.size(1)  # we only need to compute the logits for the completion tokens
 
         with torch.inference_mode():
+            old_per_token_logps = self._get_per_token_logps(
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:734]: Indeed, that's why this version is slightly slower with num_iterations=1:
W B Chart 20_02_2025, 13_24_19

Let me add this optimization

**Comment by kashif** [comment]
clearer when one realizes that the inters are the indices of the prompts!

**Review by kashif — APPROVED** [review]

**Comment by OedoSoldier** [comment]
Isn't this causes a loop?
https://github.com/huggingface/trl/blob/e5ae703d352b29537159180087ef8bd4b41bf625/trl/trainer/grpo_trainer.py#L663
https://github.com/huggingface/trl/blob/e5ae703d352b29537159180087ef8bd4b41bf625/trl/trainer/grpo_trainer.py#L679

**Comment by alexxchen** [comment]
@qgallouedec  I think we should consider adding another term to calculate $$KL(\pi_{\theta}, \pi_{\theta_{old}})$$ , because this is the KL term that effectively constrains offline updates, whereas, $$KL(\pi_{\theta}, \pi_{ref})$$ is only used to limit the overall model update. Alternatively, should we set sync_ref_model=True by default and set ref_model_sync_steps=1

**Comment by qgallouedec** [comment]
Not sure to get why you want to sync the ref model every step. It is equivalent to have no ref model, the KL will be always 0

**Comment by alexxchen** [comment]
> Not sure to get why you want to sync the ref model every step. It is equivalent to have no ref model, the KL will be always 0

Sorry, my mistake. I meant that setting ref_model_sync_steps = num_iterations might help stabilize the multi-step update. However, this would result in the loss of regularization provided by the original reference model.

**Comment by Alex-Songs** [comment]
@qgallouedec Can num_iterations not be used together with num_train_epochs? I found that setting num_train_epochs=2, the larger the num_iterations, the more iterations there are and the slower the training.

**Comment by SonuDixit** [comment]
Hi Team,
[Here](https://github.com/huggingface/trl/blob/c5ba70d4fc7cfc2e4515654eb03332d0ff4704d9/trl/trainer/grpo_trainer.py#L571), num_iterations($\mu$) optimisation steps are made on the same completion(trajectory). Please see the attached image from PPO paper - 
Screenshot 2025-03-31 at 6 15 44 PM

Attaching algorithm from DeepseekMath paper - 
Screenshot 2025-03-31 at 6 17 50 PM
I think, for line 11 in the above Algorithm - a different batch size $B$ is used such that $\mu * B \approx |D_b|$. A total of $\mu$ updates are being done using the batch size $B$. 

Please see the PPO algorithm - 
Screenshot 2025-03-31 at 6 44 42 PM
I think $k$ in the PPO-Algo corresponds to $\mu$ in the GRPO. PPO Authors have mentioned hyperparams in appendix A, wherein it can be seen that K*M is less than or approximately equal to NT.

**Comment by SonuDixit** [comment]
> Thanks @SonuDixit can you open a dedicated issue for this? Just copy-paste you comment

https://github.com/huggingface/trl/issues/3197
